accepted/0000-argument-option-parsing.md

```diff
@@ -0,0 +1,74 @@
+
+## Summary
+
+Currently, option, and argument handling in npm is not using any kind of
+standard library (either internal or a third party lib). This means that
+invalid arguments or flags are not detected.
+
+## Motivation
+
+Originally proposed in [this issue](https://github.com/npm/npm/issues/20856).
+
+Because there's no error thrown when invalid command-line flags are passed, you
+can pass in invalid flags and not realize they're invalid for a particular
+context. e.g., if you type: `npm audit --format=json` instead of `npm audit
+--json` you get the normal output of `npm audit` without any arguments, vs. an
+error that you passed an invalid flag. This should probably be fixed regardless
+of the outcome of this discussion / RFC.
+
+## Detailed Explanation
+
+This would pull in a third party lib, and / or create new methods in the core
+libs that would be used to declare subcommands, arguments, and options in a
+standardized way. Ideally, it should be self-documenting (so that usage text,
+auto-completion, help pages, etc. could be as auto-generated as possible).
+
+Then, npm itself, as well as its subcommands, could be cut over to use the new
+lib, either all at once or, more likely, one at a time, and eventually, older
+utilities could be deprecated or removed.
+
+## Rationale and Alternatives
+
+The main rationale is that exiting non-0 and throwing an error is the correct
+behavior when the user passes in ambiguous or invalid input. 
+
+Additionally, if the command line flags and args are built and documented
+using a standard library, there's less chance for drift between usage / docs
+and reality as new subcommands, flags, and arguments are added. After the
+initial work, this should also make implementing new options easier.
+
+I'm guessing it could also make keeping completion up to date easier.
+
+## Implementation
+
+This would probably be handled in `lib/config/core.js` or possible
+`lib/config/defaults.js`.
+
+This would be considered a major / breaking change.
+
+## Prior Art
+
+* [ava](https://github.com/avajs/ava/blob/master/lib/cli.js) - uses [meow](https://www.npmjs.com/package/meow)
+* [forever]() - seems to use a [custom internal lib](https://github.com/foreverjs/forever/blob/master/lib/forever/cli.js)
+
+## Unresolved Questions and Bikeshedding
+
+I haven't worked on npm before, and don't know the codebase well so I'm
+probably not the best person to determine how feasible this is or what the
+best option(s) would be.
+
+I don't know what specific libs would be best if something isn't written
+within utils  -- I'm not a JS / node dev, but it would seem like tools like
+[minimist](https://www.npmjs.com/package/minimist) or
+[commander](https://www.npmjs.com/package/commander) have few external deps,
+and might make sense to use for implementation. There are also some heavier
+weight options (e.g., [vorpal](https://vorpal.js.org/), but those might be
+tough to fit into an existing tool.
+
```

Comment:
Unfortunately, npm can't really use an external options parser :( we have some pretty special behaviors and breaking backwards compatibility with that would be a pretty risky change. I handwave this because the config system predates me by many years and I literally don't even know what all the funny little parsing cases are but I assume some of them are pretty thingy.

Reply:
I wish pretty badly that we could use [`yargs`](https://npm.im/yargs), which is probably the one that works closest to npm and is pretty swanky.